Sources/NIOCore/AsyncSequences/NIOAsyncWriter.swift

```diff
@@ -483,85 +451,42 @@ extension NIOAsyncWriter {
-
-                switch action {
-                case .callDidYield(let delegate):
-                    // We are allocating a new Deque for every write here
-                    self._lock.unlock()
-                    delegate.didYield(contentsOf: Deque(sequence))
-                    self.unbufferQueuedEvents()
-
-                case .returnNormally:
-                    self._lock.unlock()
+            while true {
+                switch try await self._yield(contentsOf: sequence, yieldID: yieldID) {
+                case .retry:
+                    continue
+                case .yielded:
                     return
-
-                case .throwError(let error):
-                    self._lock.unlock()
-                    throw error
-
-                case .suspendTask:
-                    try await withCheckedThrowingContinuation { (continuation: CheckedContinuation<Void, Error>) in
-                        self._stateMachine.yield(
-                            contentsOf: sequence,
-                            continuation: continuation,
-                            yieldID: yieldID
-                        )
-
-                        self._lock.unlock()
-                    }
-                }
-            } onCancel: {
-                // We must not resume the continuation while holding the lock
-                // because it can deadlock in combination with the underlying ulock
-                // in cases where we race with a cancellation handler
-                let action = self._lock.withLock {
-                    self._stateMachine.cancel(yieldID: yieldID)
-                }
-
-                switch action {
-                case .resumeContinuation(let continuation):
-                    continuation.resume()
-
-                case .none:
-                    break
                 }
             }
         }
 
         @inlinable
-        /* fileprivate */ internal func yield(_ element: Element) async throws {
-            let yieldID = self._yieldIDGenerator.generateUniqueYieldID()
+        /* fileprivate */ internal func _yield<S: Sequence>(contentsOf sequence: S, yieldID: StateMachine.YieldID?) async throws -> StateMachine.YieldResult where S.Element == Element {
+            let yieldID = yieldID ?? self._yieldIDGenerator.generateUniqueYieldID()
 
-            try await withTaskCancellationHandler {
+            return try await withTaskCancellationHandler {
                 // We are manually locking here to hold the lock across the withCheckedContinuation call
                 self._lock.lock()
 
-                let action = self._stateMachine.yield(contentsOf: CollectionOfOne(element), yieldID: yieldID)
+                let action = self._stateMachine.yield(contentsOf: sequence, yieldID: yieldID)
 
                 switch action {
                 case .callDidYield(let delegate):
+                    // We are allocating a new Deque for every write here
```

Comment:
Ouch. I didn't realise we did this.

Reply:
Yeah this one of the early design choices because users can yield us any kind of sequence and we have to normalise this in the `Deque` case. Right now we have provided a fast path for a single element we ought to be able to do the same for `Deque`/`Array` and more.